### PR TITLE
chore: add pull-request check to ensure plugin sidecar has memory and CPU limitations

### PR DESCRIPTION
**Pull request by vitaliy-guliy**
Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>

### What does this PR do?
Introduces a PR-check to prevent adding the plugins without or with wrong CPU and memory requests/limitations.

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/16685

### How to test this PR?

- clone this project, switch to `resource-limit-pr-check` branch
- go to `tools/automation`
- build with `yarn prepare`
- check plugins by `yarn node ./lib/check-resource-limits.js`
- change / remove cpu/memory limits/requests in `che-theia-plugins.yaml` and check plugins again

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Review by sunix — COMMENTED** [review]
To me these checks should be done in https://github.com/eclipse-che/che-plugin-registry/tree/main/tools/build
It should be one of the checks that are done there. So it would fail whether you are invoking this from your dev environment or gh-action default PR check.

**Comment by vitaliy-guliy** [comment]
> To me these check should be done in https://github.com/eclipse-che/che-plugin-registry/tree/main/tools/build
> It should be one of the checks that are done there. So it would fail whether you are invoking this from your dev environment or gh-action default PR check.

@benoitf is it OK to have PR check as it is, or is it better to rework it and make it a part of build tool?

File: .github/workflows/resource-limits-pr-check.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#
+# Copyright (c) 2020-2021 Red Hat, Inc.
```
Comment (benoitf) [inline .github/workflows/resource-limits-pr-check.yaml:2]:
```suggestion
# Copyright (c) 2021 Red Hat, Inc.
```

File: tools/automation/src/che-theia-plugins.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
+/********************************************************************************
+ * Copyright (c) 2020-2021 Red Hat, Inc.
```
Comment (benoitf) [inline tools/automation/src/che-theia-plugins.ts:2]:
```suggestion
 * Copyright (c) 2021 Red Hat, Inc.
```

File: tools/automation/src/check-resource-limits.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+/********************************************************************************
+ * Copyright (c) 2020-2021 Red Hat, Inc.
```
Comment (benoitf) [inline tools/automation/src/check-resource-limits.ts:2]:
```suggestion
 * Copyright (c) 2021 Red Hat, Inc.
```

File: tools/automation/src/resource-limits-checker.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
+/********************************************************************************
+ * Copyright (c) 2020-2021 Red Hat, Inc.
```
Comment (benoitf) [inline tools/automation/src/resource-limits-checker.ts:2]:
```suggestion
 * Copyright (c) 2021 Red Hat, Inc.
```

File: tools/automation/tests/check-resource-limits-checker.spec.ts [hunk 5]
```diff
@@ -0,0 +1,253 @@
+/**********************************************************************
+ * Copyright (c) 2020-2021 Red Hat, Inc.
```
Comment (benoitf) [inline tools/automation/tests/check-resource-limits-checker.spec.ts:2]:
```suggestion
 * Copyright (c) 2021 Red Hat, Inc.
```

**Review by benoitf — COMMENTED** [review]
Hello, I agree with Sun that we should have feedback as eary as possible.
Also for downstreaming, we might want to have the same checks as well and it won't be possible without duplicating code if it's in tools/automation while the build tool is being used everywhere (developer and PR checks)

**Comment by benoitf** [comment]
@vitaliy-guliy maybe a follow-up PR like in a step 2. (job is already done for this step)

**Comment by vitaliy-guliy** [comment]
> @vitaliy-guliy maybe a follow-up PR like in a step 2. (job is already done for this step)

I think it would be as improvement, introduced in a step 2.

**Comment by vitaliy-guliy** [comment]
Is there any volunteer to approve this PR?

**Review by benoitf — APPROVED** [review]

**Pull request merged**